R/RLearner_classif_mxff.R

```diff
@@ -0,0 +1,137 @@
+      makeNumericLearnerParam(id = "wd", default = 0,
+        requires = quote(optimizer %in% c("adadelta", "adagrad", "adam", "rmsprop", "sgd"))),
+      makeNumericLearnerParam(id = "rescale.grad", default = 1,
+        requires = quote(optimizer %in% c("adadelta", "adagrad", "adam", "rmsprop", "sgd"))),
+      makeNumericLearnerParam(id = "clip_gradient",
+        requires = quote(optimizer %in% c("adadelta", "adagrad", "adam", "rmsprop", "sgd"))),
+      makeFunctionLearnerParam(id = "lr_scheduler",
+        requires = quote(optimizer %in% c("adagrad", "adam", "rmsprop", "sgd"))),
+      makeNumericLearnerParam(id = "learning.rate",
+        requires = quote(optimizer %in% c("adagrad", "adam", "rmsprop", "sgd"))),
+      makeNumericLearnerParam(id = "beta1", default = 0.9, requires = quote(optimizer == "adam")),
+      makeNumericLearnerParam(id = "beta2", default = 0.999, requires = quote(optimizer == "adam")),
+      makeNumericLearnerParam(id = "gamma1", default = 0.95,
+        requires = quote(optimizer == "rmsprop")),
+      makeNumericLearnerParam(id = "gamma2", default = 0.9,
+        requires = quote(optimizer == "rmsprop")),
+      makeNumericLearnerParam(id = "momentum", default = 0, requires = quote(optimizer == "sgd"))
+    ),
+    properties = c("twoclass", "multiclass", "numerics"),
+    par.vals = list(learning.rate = 0.1, array.layout = "rowmajor", verbose = FALSE),
+    name = "Feedforward Neural Network",
+    short.name = "mxff",
+    note = "Default of `learning.rate` set to `0.1`. Default of `array.layout` set to `'rowmajor'`.
+    Default of `verbose` is set to `FALSE`. If `symbol` is specified, it will be passed to mxnet
+    ignoring other architectural specifications. Default of `initializer` is set to NULL, which
+    results in the default mxnet initializer being called when training a model. Number of output
+    nodes is detected automatically."
+  )
+}
+
+#' @export
+trainLearner.classif.mxff = function(.learner, .task, .subset, .weights = NULL,
+  layers = 1L, nodes1 = 1L, nodes2 = NULL, nodes3 = NULL, nodes.out = NULL,
+  act1 = "tanh", act2 = NULL, act3 = NULL, act_out = "softmax", dropout = NULL, symbol = NULL,
+  ...) {
+  # transform data in correct format
+  d = getTaskData(.task, subset = .subset, target.extra = TRUE)
+  y = as.numeric(d$target) - 1
+  X = data.matrix(d$data)
+
+  # construct vectors with #nodes and activations
+  if (!is.null(symbol)) {
+    out = symbol
+  } else {
+    sym = mx.symbol.Variable("data")
+    act = list(act1, act2, act3)[1:layers]
+    nodes = list(nodes1, nodes2, nodes3)[1:layers]
+
+    # construct hidden layers using symbols
+    for (i in seq_len(layers)) {
+      sym = mx.symbol.FullyConnected(sym, num_hidden = nodes[[i]])
+      sym = mx.symbol.Activation(sym, act_type = act[[i]])
+    }
+
+    # add dropout if specified
+    if (!is.null(dropout)) {
+      dropout = min(dropout, (1 - 1e-7))
```

Comment:
`dropout` needs to be in [0,1) and the `mxnet` function `mx.mlp()` which also implements this had it like that.
But you're absolutely right, this should be made explicit. I changed the upper bound for `dropout` and also commented on it in the notes.